### ISSUE TITLE
[Feature Request

### ISSUE THREAD

**Issue by EricJ2190** [simplex-chat/simplex-chat#1081]

**Comment by EricJ2190** [comment]
UnifiedPush allows the user to choose a "distributor" to handle the push delivery, and the security would depend on that choice. I would expect that SimpleX would put minimal information in the actual push message and just use it as a cue to fetch data from the message queue server, similar to how push is handled on iOS. A malicious distributor could prevent you from getting immediately notified about new messages but could not view, manipulate, or censor messages.

**Comment by epoberezkin** [comment]
Our code for receiving messages follows exactly the same approach as ntfy, so you would only have benefits with ntfy if you have multiple apps using the same channel.

We need to systematically eliminate any sources of unnecessary traffic (e.g. too many retires, in cases where the failure would be more appropriate) - it is likely to have more positive impact on battery usage than switching to ntfy.

**Comment by ThatOneCalculator** [comment]
I have 5 apps using NTFY on my phone, so UnifiedPush would be very much appreciated

**Comment by mdebus** [comment]
[https://f-droid.org/en/2022/12/18/unifiedpush.html](Here) is a good explanation of UnifiedPush:

@epoberezkin Would It be even be possible to combine your approach with UnifiedPush?

**Comment by mrg358** [comment]
Big plus for ntfy as it can be easily selfhosted.
FWIW popular matrix clients (fluffy chat and element) are automatically utilizing `ntfy` if it is detected running on a users device.

**Comment by ltGuillaume** [comment]
@mrg358 Conversations can be used by other apps for UnifiedPush as well.

**Comment by khimaros** [comment]
> Our code for receiving messages follows exactly the same approach as ntfy, so you would only have benefits with ntfy if you have multiple apps using the same channel.
> 
> We need to systematically eliminate any sources of unnecessary traffic (e.g. too many retires, in cases where the failure would be more appropriate) - it is likely to have more positive impact on battery usage than switching to ntfy.

I also have six apps on my ntfy right now. with simplex that would become seven. instead I'm using periodic notifications to reduce battery drain.

**Comment by iroedius** [comment]
I use NextPush (which is a UnifiedPush distributor) and I would also appreciate it if it was supported.

**Comment by KaiserCalm** [comment]
I've recently discovered ntfy and on a degoogled phone it's really awesome. Would be a great feature!

**Comment by Darin755** [comment]
I would really like to see this be implemented. I use unifiedpush with several other apps using ntfy and its very efficient and allows for a single app to handle everything

**Comment by ThatOneCalculator** [comment]
It seems like more and more apps are adding UnifiedPush support (including Molly, which is big!), so getting more privacy-respecting messaging apps like Simplex to join the likes of Element/FluffyChat/SchildiChat/SmallTalk for Matrix, Molly for Signal, Goguma for IRC, Tox PMA for Tox, Jami, and Databag would lead to further adoption of UnifiedPush to provide a *unified* way for push notifications to be handled without being handled by a data-hoarding company.

**Comment by gvansanden** [comment]
> Our code for receiving messages follows exactly the same approach as ntfy, so you would only have benefits with ntfy if you have multiple apps using the same channel.
> 
> We need to systematically eliminate any sources of unnecessary traffic (e.g. too many retires, in cases where the failure would be more appropriate) - it is likely to have more positive impact on battery usage than switching to ntfy.

Next push on my phone receives a lot of notifications during a typical day and uses less then 1% of battery.

Simplex receives at most 2 notifications a day and uses 20% battery. Which is dramatic on my pixel 7 which already doesn't have the best battery life

**Comment by jm355** [comment]
Just chiming in to say I have 5 apps using ntfy for notifications, and 2 topics I subscribe to (status updates for different automated tasks on my server). It would be awesome if simplex could just take advantage of that, and consolidate to one service. I self host ntfy as well. Molly has it and it works great, very minimal battery drain

**Comment by hstock** [comment]
I would like to see this as well. As far as I have seen the implementation guidelines, the notification can wake up the App, so the notification itself does not need to show content? https://unifiedpush.org/developers/android/

The app received the full POST body, so even encrypted content could be in there. 

For self-hosted servers, the privacy implications would have to be analyzed, though.

**Comment by pm4rcin** [comment]
I agree that best case would be UnifiedPush. If not then something like tutanota did for their android app is also not bad from battery perspective. https://f-droid.org/2018/09/03/replacing-gcm-in-tutanota.html

**Comment by ThatOneCalculator** [comment]
Please keep in mind that article is from 2018 and Tuta now recommends UnifiedPush. https://tuta.com/blog/google-push-alternative/

**Comment by robsmith11** [comment]
Lack of push notifications is the main issue keeping some of my non-tech friends from switching to SimpleX.

Even if there is some sacrifice in privacy, I think (optional) push support would be a big win for adoption by more casual users.

**Comment by gvansanden** [comment]
> Lack of push notifications is the main issue keeping some of my non-tech friends from switching to SimpleX.
> 
> Even if there is some sacrifice in privacy, I think (optional) push support would be a big win for adoption by more casual users.

If the instant option would not drain battery, it would negate the need for push.
The problem with push notifications for non-techies is that it would require the inclusion of Google libraries which contain trackers

**Comment by jm355** [comment]
> The problem with push notifications for non-techies is that it would require the inclusion of Google libraries which contain trackers

Setting up unified-push is super easy. You can just have them install ntfy and it should just work

**Comment by gvansanden** [comment]
> > The problem with push notifications for non-techies is that it would require the inclusion of Google libraries which contain trackers
> 
> Setting up unified-push is super easy. You can just have them install ntfy and it should just work

Yes and no, you need to have it running on a server or use one provided by someone.  Secondly, we are talking about non-technical users

**Comment by ThatOneCalculator** [comment]
I told my parents over text how to setup unifiedpush (and they're very non-technical people), and they were able to get it working with no problem. The trick? Just download the ntfy app and follow its setup instructions.

**Comment by erebion** [comment]
How to set up ntfy:

1. Open the app

Wow, that was hard.

**Comment by erebion** [comment]
(I mean, open it once.)

**Comment by EricJ2190** [comment]
UnifiedPush has two optional embedded FCM distributors available, one (in beta) that is FOSS. This would allow users with Google services installed to be able to use it for push notifications if they wish, while allowing anyone else to use the UnifiedPush distributor of their choice (like ntfy) to handle push notifications.

I think this would avoid concerns about battery and reliability for casual users, while also providing flexibility for anyone who wants it.

**Comment by m1chj** [comment]
@epoberezkin gonna reconsider? You received quite a lot of constructive feedback

**Comment by jm355** [comment]
I'm going to uninstall simplex until this is added. Molly has it and my group all uses signal, so unless simplex gets this then I would be losing features if we switched, so unless there's some big privacy benefit then it's not worth it.

**Comment by prkuctd** [comment]
+1 for the feature

**Comment by imgalaxyhunter** [comment]
if someone were to make a fork with UnifiedPush support, would you merge it?

**Comment by imgalaxyhunter** [comment]
> [https://f-droid.org/en/2022/12/18/unifiedpush.html](Here) is a good explanation of UnifiedPush:
> 
> @epoberezkin Would It be even be possible to combine your approach with UnifiedPush?

from my observations, yes, iOS app optionally uses push servers as there's no other option for instant push notifications. it should be fairly easy for someone to implement UnifiedPush using the same technique.

**Comment by SKBotNL** [comment]
Even though v6 does seem to have reduced battery usage for me, I´d, just like many other users, would like to request UnifiedPush support

**Comment by pm4rcin** [comment]
Looks like that this might happen eventually but I guess not soon. Evgeny's response on reddit:
![obraz](https://github.com/user-attachments/assets/f3fe1af5-3042-44ee-beef-0681de3c2423)

**Comment by jm355** [comment]
Might happen once the current approach is optimized?

The whole point of unified push is that the current approach doesn't need to be optimized...

**Comment by m6n50u7** [comment]
as far as I understand the way simpleX works, this will defeat all security purpose.
unless you'd run a special external server that'd act as client and then notify you through unified push when there is new message.

**Comment by jm355** [comment]
No, all that would get sent via unifiedpush is a thing telling the app there might be new messages, so it can wake up and check. That's how molly-unifiedpush does it, no details of any conversations get sent over unifiedpush

**Comment by jm355** [comment]
https://github.com/simplex-chat/simplex-chat/issues/1081#issuecomment-1254272487 @m6n50u7

**Comment by jtrees** [comment]
> this will defeat all security purpose. unless you'd run a special external server

That's actually a common use case. I run my own ntfy server for push notifications.

**Comment by m6n50u7** [comment]
it doesn't have to be containing the message. there is the ip protection that won't be active while the push app checks for notifications, and corelating the different chats for one user.

**Comment by m6n50u7** [comment]
> I'm not sure what you mean. The push app is regularly checking for notifications (i don't know the exact implementation, it depends on the UP distributor being used), and provides notification support for any apps compatible with unifiedpush. Simplex would just get a hint from the push app that there may or may not be new notifications, and would then wake up and check for notifications in the same way as it fetches notifications now, only instead of checking at certain time intervals, it would only check when it gets that hint.

the way push works is:
the server sends a push request to your push provider to tell you there is new message.
this is easy in apps like molly because signal's server knows your account and people usually use it with no relays/proxies.
however, in simpleX your account is only stored locally and thus all the server knows is that there is bunch of channels which may be used for anything. and on top of that, simpleX would be mostly connecting with the IP protection.
in order to directly use UP these two security features would have to be canceled. because when UP connects to the UP server it won't use IP protection and the provider would be able to notice all these different channels belongs to one user. which means your account will be kind of known.

**Comment by m6n50u7** [comment]
however, there is a possible workaround which is that there will be another simpleX app which works in this way:
android's client will be just a remote control. while the other client will be doing the real stuff on your own server.
whenever there is new message the simpleX client on your server will send you a push notification (preferably self hosted) and wait till your app fetch the messages from it.

**Comment by EricJ2190** [comment]
It would likely just work like the iOS method described here:
https://simplex.chat/blog/20220404-simplex-chat-instant-notifications.html

Each contact would have a separate queue for sending notifications to a notification server. The notification server would then send the notification onto the UnifiedPush endpoint. The UnifiedPush provider would not see any information about SMP queues or the sender. It will just be able to see that receiving device uses SimpleX and is receiving messages.

**Comment by iopq** [comment]
This is my #1 issue on Android, I'm currently using SchildiChat with ntfy and a few more

**Comment by theotheroracle** [comment]
this is the biggest reason holding me back from using simplex currently

**Comment by SinnySupernova** [comment]
It's clear that the topic is important for many people 🚀  

Is there a way for those who aren't knowledgeable in Haskell to contribute to this or somehow push it forward in the priority queue?

**Comment by gvansanden** [comment]
@SinnySupernova The thing is that even though I initially very much was in favor of UnifiedPush, getting to know the inner workings of Simplex it would indeed greatly reduce anonimity.   As SimpleX doesn't work much different from unifiied push and based their implementation on ntfy.sh, it seems more likely that the thing preventing the phone from deep sleeping is the actual issue and not the notifications themselves.
So I  think it can be fixed by not adding unified push and staying anonymous

**Comment by jm355** [comment]
One way to do it would be as follows:

Say Bob has unifiedpush set up. When Bob messages Joe, part of the message data (which is encrypted, so only Joe can read it) contains the UP url for simplex, something like ntfy.sh/UPxyz123. That wouldn't be displayed to Joe, but it would be saved by the simplex app. Then, when Joe messages Bob, Joes phone directly pings the UP url with a generic type of "there may be new messages" which, when Bobs phone gets the notification triggers Bobs phone to wake up simplex and check for any new messages. 

But then again, that might give the UP server more info than desired, since it would see that Joes ip is notifying Bob. It sounds like this is solved by the "iOS method", where a separate server handles pinging the UP server, so the UP server only sees simplex requests from one IP?

**Comment by SinnySupernova** [comment]
I'd like to point out that the biggest reason why I want UP is to have a single way of setting things up on my phone for a number of applications. IMO that makes it less error prone and easier to keep track of. 

I don't know who sends the request to the UP server - the client device or a server that processes the message delivery. Depending on that, the concerns about the UP server knowing too much might be irrelevant.

Another question is whether self-hosting a ntfy server (or having a trusted one) would help.

**Comment by gvansanden** [comment]
@SinnySupernova I like that about UP too, but the push mechanism in Simplex requires no additional setup for you.  Having your own Ntfy server actually makes it easier to link a queue with your server, which is your identity and I think that is why the devs don't want to use it (instead rebuilding the ntfy mechanism).

That said, since someone identified that SimpleX keeps the phone awake, the battery drain seems to be a seperate issue from UP support, if that is the case, the better approach is to keep the built-in notification system

**Comment by m6n50u7** [comment]
as I said, the only way to do this would require using SXC on a headless server owned by you and you control through your phone. and that server will also handle the push notifications.
any other suggested way will deanonymize the user.

**Comment by imgalaxyhunter** [comment]
It's only more private to use your own server if you're using flawed systems that don't protect your data by design. If you have a good system, like SimpleX, it's more private to use a server shared by many users rather than your own to thwart traffic analysis.

**Comment by 0cwa** [comment]
In that case perhaps SimpleX should implement its own anonymous UNP gateway? Making ntfy.sh more or less obsolete from a privacy POV 😉

**Comment by summoner001** [comment]
Can't wait for this :)

**Comment by schklom** [comment]
Hi, I just wanted to mention that I made a feature request on ntfy (https://github.com/binwiederhier/ntfy/issues/1278) to allow SimpleX to send notifications to ntfy.sh which would forward them to any self-hosted instance.

If accepted and a user wants to use their own ntfy instance, they can go in ntfy settings to define the UnifiedPush relay server (ntfy.sh), SimpleX will use that to register a topic and let ntfy handle receiving notifications.
If Alice sets up alice_ntfy.domain and wants to use ntfy.sh as a relay, Bob's SimpleX will send a notification to ntfy.sh, ntfy.sh will relay the notification to alice_ntfy.domain, and alice_ntfy.domain will pass the notification to Alice's phone.

@imgalaxyhunter I believe this would solve the privacy problem of a self-hosted server being used to identify its owner which you mentioned in https://github.com/simplex-chat/simplex-chat/issues/1081#issuecomment-2495384494, while allowing the phone to only connect to a self-hosted ntfy server.

If the feature request is accepted, the ntfy dev might be open to PRs if anyone knows how to implement this.

**Comment by kuflierl** [comment]
1+ for supporting unified push.

**Comment by 497a** [comment]
+1 for supporting UnifiedPush

**Comment by schklom** [comment]
@kuflierl @497a @balintbarna You should upvote (thumb up usually) the issue's first post instead of writing +1. Popularity/demand can be easily assessed with reactions on GitHub, but no one is going to count all the comments saying +1.

**Comment by ghost** [comment]
> Our code for receiving messages follows exactly the same approach as ntfy, so you would only have benefits with ntfy if you have multiple apps using the same channel @epoberezkin

That is exactly why I would like it if simplex implemented unified push. Because other apps on my phone use it as well. That's basically the whole point of unified push. Hence the unified.

This is the most upvoted issue on the repo. Is the dev team on board with something like this? Or does it not have priority atm?

> Very interesting, but I do not know on the security side whether someone or something can alter or cancel notifications at its will (malicious actor).
> In that sense I leave the evaluation to those who surely understand more than me.

You can do what signal does and just ping without sending message content. Or you can use e2e encryption. More information can be found here: https://unifiedpush.org/

**Comment by p1gp1g** [comment]
FYI, I'm working on it

**Comment by gvansanden** [comment]
To re-iterate.  Despite my earlier support it wil impact anonimity on Simplex, but I've also don't believe it will solve the problem.  Signal or Matrix in polling mode use 1-8% of my Pixel's battery.  Even if I don't add any contacts or groups, SimpleX will eat 30-40%.  
This indicates to wakelocks as a cause more then the actual polling itself and that won't go away with unified push either

**Comment by ghost** [comment]
To clarify my point: the point of integrating something like unified push is not to resolve any battery issues that are caused by simplex code. The point is to allow using a single connection to the unified push service for _all_ push notifications vs. each app implementing their own.

Users should of course always be free to choose to use the simplex connection if they prefer to do so. But, there are some who prefer using unified push because it requires just the single connection. Personally I just don't want any app's background service active for notifications other than my unified push connection.

**Comment by p1gp1g** [comment]
I've opened the PRs

**Comment by ghost** [comment]
I was wondering how things are going with the unifiedpush proposal. Were there any roadblocks?

**Comment by AbdulrahmanNahhas** [comment]
+1 for supporting UnifiedPush

**Comment by schklom** [comment]
@AbdulrahmanNahhas You should upvote (thumb up usually) the issue's first post instead of writing +1. Popularity/demand can be easily assessed with reactions on GitHub, but no one is going to count all the comments saying +1.

**Comment by AbdulrahmanNahhas** [comment]
> @AbdulrahmanNahhas You should upvote (thumb up usually) the issue's first post instead of writing +1. Popularity/demand can be easily assessed with reactions on GitHub, but no one is going to count all the comments saying +1.

🫡

**Comment by Yiannis128** [comment]
I think unified push is a must for making SimpleX daily drivable.

**Comment by CherrieTheShifter** [comment]
Yes This

**Comment by MrMichalskiProjects** [comment]
Please add the unified push, I beg you, it's the only thing keeping me form migrating my normie friends from discord to this masterpiece of an app.

**Comment by atiq-cs** [comment]
Unified Push is the future. To compete with matrix, must have to implement Unified Push.

**Comment by dessalines** [comment]
I've been comparing simpleX battery usage vs UPNP apps over the last two weeks, and it's not even comparable. SimpleX uses 4x the background battery that they do.

Not only should Upnp be added as the highest priority for this app, the previous background service be completely replaced by it.
